compiler/nilcheck.nim

```diff
@@ -0,0 +1,624 @@
+  map.locals[symbol] = value
+  map.history.mgetOrPut(symbol, @[]).add(History(info: info, kind: kind, nilability: value))
+
+proc hasKey(map: NilMap, name: Symbol): bool =
+  var now = map
+  result = false
+  while not now.isNil:
+    if now.locals.hasKey(name):
+      return true
+    now = now.previous
+
+iterator pairs(map: NilMap): (Symbol, Nilability) =
+  var now = map
+  while not now.isNil:
+    for name, value in now.locals:
+      yield (name, value)
+    now = now.previous
+
+proc copyMap(map: NilMap): NilMap =
+  if map.isNil:
+    return nil
+  result = newNilMap(map.previous.copyMap())
+  for name, value in map.locals:
+    result.locals[name] = value
+  for name, value in map.history:
+    result.history[name] = value
+
+proc `$`(map: NilMap): string =
+  var now = map
+  var stack: seq[NilMap] = @[]
+  while not now.isNil:
+    stack.add(now)
+    now = now.previous
+  for i in countdown(stack.len - 1, 0):
+    now = stack[i]
+    result.add("###\n")
+    for name, value in now.locals:
+      result.add(&"  {name} {value}\n")
+  
+proc symbol(n: PNode): Symbol =
+  aecho n, n.kind
+  case n.kind:
+  of nkSym:
+    if $n == "result":
+      result = resultId
+    else:
+      result = n.sym.id
+  of nkHiddenAddr, nkAddr:
+    result = symbol(n[0])
+  else:
+    result = hashTree(n)
+  aecho result
+
+using
+  n: PNode
+  conf: ConfigRef
+  map: NilMap
+
+proc typeNilability(typ: PType): Nilability
+
+
+
+proc checkCall(n, conf, map): Check =
+  var isNew = false
+  result.map = map
+  for i, child in n:
+    discard check(child, conf, map)
+    if i > 0 and child.kind == nkHiddenAddr:
+      if child.typ.kind == tyVar and child.typ[0].kind == tyRef:
+        # yes
+        if not isNew:
+          result.map = newNilMap(map)
+          isNew = true
+        # result.map[$child] = MaybeNil
+        # echo "MaybeNil arg"
+        # echo "  ", child
+        # echo "  ", symbol(child)
+        let a = symbol(child)
+        result.map.store(a, MaybeNil, TVarArg, n.info)
+        # echo result.map
+
+  if n[0].kind == nkSym and n[0].sym.magic == mNew:
+    # let b = $n[1]
+    let b = symbol(n[1])
+    # result.map[b] = Safe
+    result.map.store(b, Safe, TAssign, n[1].info)
+    result.nilability = Safe
+  else:
+    result.nilability = typeNilability(n.typ)
+  # echo result.map
+
+proc event(b: History): string =
+  case b.kind:
+  of TArg: "param with nilable type"
+  of TNil: "isNil"
+  of TAssign: "assigns a value which might be nil"
+  of TVarArg: "passes it as a var arg which might change to nil"
+  of TResult: "it is nil by default"
+  of TType: "it has ref type"
+
+proc derefWarning(n, conf, map; maybe: bool) =
+  var a = history(map, symbol(n))
+  var res = ""
+  res.add("can't deref " & $n & " it " & (if maybe: "might be" else: "is") & " nil")
+  res.add("\n")
+  for b in a:
+    res.add("  " & event(b) & " on line " & $b.info.line & ":" & $b.info.col)
+  message(conf, n.info, warnNilCheck, res)
```

Comment:
```suggestion
  var res = @["can't deref " & $n & " it " & (if maybe: "might be" else: "is") & " nil"]
  for b in a:
    res.add "  " & event(b) & " on line " & $b.info.line & ":" & $b.info.col
  message(conf, n.info, warnNilCheck, res.join"\n")
```
Basically `.add` to 1 `seq[string]` then `.join"\n"` when returning.

Reply:
good idea! but maybe allocations ? it is a hotspot